### PR TITLE
Add RIOSocketWrapper and SaeaSocketWrapper

### PR DESCRIPTION
**Pull request by hebinhuang**
Run Sparkclrsperf.exe for perf suite RunRDDMaxDeletionsByUser only (because this suite invoke CSharpWorker).

There is 11% perf improvement by using RIO socket for the file with 4KB a line, and 15% perf improvement for the file with 64KB a line.

Traditional Socket for file with 4KB a line:
User with max deletions is /user/mwcl_musicbrainz, count of deletions=237776. Elapsed time=00:04:35.1505607
Requesting to close all call back sockets.
** Printing results of the perf run (C#) **
** Execution time for RunRDDMaxDeletionsByUser in Milliseconds. Min=275150, Max=275150, Average=275150, Median=275150, Number of runs=1, Individual execution duration=[275150] **
** *** **

Rio Socket for file with 4KB a line:
User with max deletions is /user/mwcl_musicbrainz, count of deletions=237776. Elapsed time=00:04:07.2763483
Requesting to close all call back sockets.
** Printing results of the perf run (C#) **
** Execution time for RunRDDMaxDeletionsByUser in Milliseconds. Min=247276, Max=247276, Average=247276, Median=247276, Number of runs=1, Individual execution duration=[247276] **
** *** **

Traditional Socket for file with 64KB a line:
User with max deletions is /user/mwcl_musicbrainz, count of deletions=5167. Elapsed time=00:01:47.3907642
** Printing results of the perf run (C#) **
** Execution time for RunRDDMaxDeletionsByUser in Milliseconds. Min=107390, Max=107390, Average=107390, Median=107390, Number of runs=1, Individual execution duration=[107390] **
** *** **

Rio Socket for file with 64KB a line:
User with max deletions is /user/mwcl_musicbrainz, count of deletions=5167. Elapsed time=00:01:33.1928019
** Printing results of the perf run (C#) **
** Execution time for RunRDDMaxDeletionsByUser in Milliseconds. Min=93192, Max=93192, Average=93192, Median=93192, Number of runs=1, Individual execution duration=[93192] **
** *** **

Perf results for RIO socket by SocketTest.exe(No spark):
send/recv data without operation: 164.4418838 seconds/10000 messages (10 MB per message)
send/recv data with operation: 1026.1169171 seconds/10000 messages (10 MB per message)

Perf results for tradition socket by SocketTest.exe(No spark):
send/recv data without operation: 1159.0978946 seconds/10000 messages (10 MB per message)
send/recv data with operation: 2555.039326 seconds/10000 messages (10 MB per message)

### CODE REVIEW

**Comment by skaarthik** [comment]
@hebinhuang - the numbers are good to see. How real-world-like is the data you used for benchmarking? I am wondering what % of typical workloads will benefit from this improvement. I am aware that it may not be easy to answer that question but that estimate could help us decide whether to keep the traditional socket as the default or not.

**Comment by hebinhuang** [comment]
I have confirmed with Renyi this afternoon. The size of each line is about 4 ~ 5 KB in our internal project. So the file with 4KB a line is close to real-world-like, which has 11% improvement.

File: csharp/build.sh [hunk 1]
```diff
@@ -37,7 +37,7 @@ export CONFIGURATION=$STEP
 
 export STEP=$CONFIGURATION
 
-xbuild /p:Configuration=$CONFIGURATION $XBUILDOPT $PROJ
+xbuild /p:Configuration=$CONFIGURATION;AllowUnsafeBlocks=true $XBUILDOPT $PROJ
```
Comment (qualiu) [inline csharp/build.sh]: "xbuild /p:Configuration=$CONFIGURATION;AllowUnsafeBlocks=true" can be error on Linux as the semicolon ";" is treated as the separator of two commands.
That's why compilation failed : [./build.sh: line 40: /verbosity:minimal: No such file or directory]( https://travis-ci.org/Microsoft/Mobius/builds/139925856#L2208)

File: csharp/build.sh [hunk 2]
```diff
@@ -46,7 +46,7 @@ export STEP=Release
 
 export CONFIGURATION=$STEP
 
-xbuild /p:Configuration=$CONFIGURATION $XBUILDOPT $PROJ
+xbuild /p:Configuration=$CONFIGURATION;AllowUnsafeBlocks=true $XBUILDOPT $PROJ
```
Comment (qualiu) [inline csharp/build.sh]: And this line is also need to be corrected with the same reason of line 40.

File: csharp/build.sh (hunk 1 above)
Comment (hebinhuang) [inline csharp/build.sh]: @qualiu, does it use "," comma?

Comment (qualiu) [inline csharp/build.sh]: Probably you can quote them just like : 
xbuild "/p:Configuration=$CONFIGURATION;AllowUnsafeBlocks=true"

Wait a minute, I'll restart my computer to verify that on virtual CentOS.

Comment (qualiu) [inline csharp/build.sh]: Verified OK on CentOS 7, just add quotes: (sorry take a long time to prepare the env)
xbuild "/p:Configuration=$CONFIGURATION;AllowUnsafeBlocks=true" $XBUILDOPT $PROJ

Comment (hebinhuang) [inline csharp/build.sh]: @qualiu 👍

Comment (hebinhuang) [inline csharp/build.sh]: @qualiu, is the quotation mark before /p: or following /p:?

**Comment by skaarthik** [comment]
@hebinhuang - FYI: I cannot build the local clone of your repo in my dev box. It fails because of the missing native build toolset. We need to make the build experience and first-time experience smooth for .NET devs who may not have toolset for building vcxproj.

File: csharp/build.sh (hunk 1 above)
Comment (qualiu) [inline csharp/build.sh]: @hebinhuang : just as the latest comment shown, 
quote the intact item as intact : "/p:Configuration=$CONFIGURATION;AllowUnsafeBlocks=true"

**Comment by hebinhuang** [comment]
@skaarthik - does it requires VC++ installed? my dev box has VC++.

**Comment by skaarthik** [comment]
@hebinhuang - yes, it looks like VC++ installation is required to build the code. Can we put checks in place to let the build exclude vcxproj if the toolset is not available in the machine? That also makes a case for making traditional socket the default option (we should make it trivially easy to switch to the optimized implementation though)

Please also review and address the code coverage gaps reported by CodeCov.

**Comment by qualiu** [comment]
@hebinhuang : LooksGoodToMe : Verified(build + run-samples) on Windows/CentOS-7.0/Ubuntu-16.04 by:
```
git clone https://github.com/Microsoft/Mobius.git
git branch hb
git checkout hb
git pull https://github.com/hebinhuang/Mobius.git ManagedRioSock
```

**Comment by hebinhuang** [comment]
Thanks @skaarthik and @qualiu for the review.
I will focus on to add more unit test for the CodeCov. For the build, I will bring a fresh box and see it how it work.

**Comment by danielli90** [comment]
Submitted rebuild at AppVeyor.

**Comment by skaarthik** [comment]
@danielli90 - do we know what caused the previous failure in AppVeyor? Let's not miss the opportunities to identify transient errors caused by the bugs in the code.

**Comment by danielli90** [comment]
AppVeyor keeps track of PR build history. The first failure is an appveyor issue - hence the restart.

**Comment by skaarthik** [comment]
Just checked the previous build error. It says "Exception encountered unloading AppDomain". It does not seem like an AppVeyor issue.

**Comment by skaarthik** [comment]
Looks the other earlier builds failed with the same error message.

**Comment by danielli90** [comment]
yes - "xception encountered unloading AppDomain". @hebinhuang is looking into it.

**Comment by hebinhuang** [comment]
It looks like the error "Exception encountered unloading AppDomain" is caused by the backend threads not terminal.
1. The unit test of SocketWrapperTest has a while(true) thread. I have fixed it.
2. The WeakObjectManagerImp in SparkCLRTestEnvironment is not explicitly disposed. So the thread in WeakObjectManagerImp is always running until GC collect it.

**Comment by hebinhuang** [comment]
Hi @skaarthik,
@danielli90 and I discussed the root cause of the error "Exception encountered unloading AppDomain". After disabled/enabled RIOSocketWrapper unit test, we have found the root cause. The free AppVeyor has only one core. Our Mobius has two backend threads that created by RioNative and WeekObjectManagerImp singleton object. The two backend threads keep running until RioNative and WeekObjectManagerImp object got disposed (process exist).
Nunit will waits for all thread terminated after all test finished. Due to the single core of AppVeyor, the two backend threads cannot be terminated in time, then Nunit hit unloading AppDomain exception.

In the fix, I explicitly calls dispose() of RioNative in the unit test. I have tried two times, and no failure.

**Comment by danielli90** [comment]
LGTM.

**Comment by skaarthik** [comment]
let's not mix maven update with other changes in this PR. I have sent a separate PR with fixes both for AppVeyor and Travis

**Comment by hebinhuang** [comment]
@skaarthik, I will revert it once your PR is merged.

**Pull request merged**